### PR TITLE
use hexchar in stdlib

### PR DESCRIPTION
**Pull request by ringabout**
(EDIT: refs https://github.com/timotheecour/Nim/issues/430)

### CODE REVIEW

File: lib/pure/parsecfg.nim [hunk 1]
```diff
@@ -248,18 +248,9 @@ proc getFilename*(c: CfgParser): string {.rtl, extern: "npc$1".} =
...
-    inc(c.bufpos)
-  of 'a'..'f':
-    xi = (xi shl 4) or (ord(c.buf[c.bufpos]) - ord('a') + 10)
-    inc(c.bufpos)
-  of 'A'..'F':
-    xi = (xi shl 4) or (ord(c.buf[c.bufpos]) - ord('A') + 10)
-    inc(c.bufpos)
-  else:
-    discard
+  var incrFlag = false
```
Comment (timotheecour) [inline lib/pure/parsecfg.nim]: can we instead change, decode_helpers.handleHexChar as follows:
```nim
proc handleHexChar*(c: char, x: var int, f: var bool) {.inline.} =
```

=>

```nim
proc handleHexChar*(c: char, x: var int): bool {.inline.} =
```

it'll simplify most usages.

**Comment by ringabout** [comment]
I have cleaned most of `hexchar` in stdlib.
Except:
- [ ] std/oids: `proc hexbyte*(hex: char): int =`
- [ ] compiler/lexer
```nim
proc handleHexChar(L: var Lexer, xi: var int; position: range[0..4])
```
- [ ] compiler/rodutils
`proc hexChar(c: char, xi: var int) =`

They will be cleaned in the future PR.

File: lib/pure/parsecfg.nim [hunk 2]
```diff
@@ -305,8 +291,10 @@ proc getEscapedChar(c: var CfgParser, tok: var Token) =
   of 'x', 'X':
     inc(c.bufpos)
     var xi = 0
-    handleHexChar(c, xi)
-    handleHexChar(c, xi)
+    if handleHexChar(c.buf[c.bufpos], xi):
+      inc(c.bufpos)
+    if handleHexChar(c.buf[c.bufpos], xi):
```
Comment (timotheecour) [inline lib/pure/parsecfg.nim]: previous code couldn't optimize that but now you can:
```nim
if handleHexChar(c.buf[c.bufpos], xi):
  inc(c.bufpos)
  if handleHexChar(c.buf[c.bufpos], xi):
    inc(c.bufpos)
```

File: lib/pure/parsesql.nim [hunk 3]
```diff
@@ -130,8 +116,10 @@ proc getEscapedChar(c: var SqlLexer, tok: var Token) =
   of 'x', 'X':
     inc(c.bufpos)
     var xi = 0
-    handleHexChar(c, xi)
-    handleHexChar(c, xi)
+    if handleHexChar(c.buf[c.bufpos], xi):
+      inc(c.bufpos)
+    if handleHexChar(c.buf[c.bufpos], xi):
```
Comment (timotheecour) [inline lib/pure/parsesql.nim]: same comment as https://github.com/nim-lang/Nim/pull/16290/files#r543656396

File: lib/pure/pegs.nim [hunk 4]
```diff
@@ -1515,8 +1503,10 @@ proc getEscapedChar(c: var PegLexer, tok: var Token) =
       tok.kind = tkInvalid
       return
     var xi = 0
-    handleHexChar(c, xi)
-    handleHexChar(c, xi)
+    if handleHexChar(c.buf[c.bufpos], xi):
+      inc(c.bufpos)
+    if handleHexChar(c.buf[c.bufpos], xi):
```
Comment (timotheecour) [inline lib/pure/pegs.nim]: see https://github.com/nim-lang/Nim/pull/16290/files#r543656396

File: lib/std/private/decode_helpers.nim [hunk 5]
```diff
@@ -1,9 +1,25 @@
...
+  ## Returns `true` if `c` is hexadecimal.
+  ##
+  ## When `c` is hexadecimal, the proc is equal to `x = x shl 4 + hex2Int(c)`.
+  runnableExamples:
+    var x: int = 0
+    assert handleHexChar('a', x)
+    assert x == 10
+
+    assert handleHexChar('B', x)
+    assert x == 171 # 10 << 4 + 11
```
Comment (timotheecour) [inline lib/std/private/decode_helpers.nim]: use nim, not c :)
```nim
assert x == 171 # (10 shl 4) + 11
```

File: lib/std/private/decode_helpers.nim [hunk 6]
```diff
@@ -1,9 +1,25 @@
...
+
+    # failed
+    assert not handleHexChar('?', x)
```
Comment (timotheecour) [inline lib/std/private/decode_helpers.nim:14]: add:
```nim
assert x == 171 # unchanged
```
(examples are part of documentation, no need to also clarify that in doc comment, it'd be redundant with this)

File: lib/std/private/decode_helpers.nim [hunk 7]
```diff
@@ -14,9 +30,9 @@ proc decodePercent*(s: openArray[char], i: var int): char =
   result = '%'
   if i+2 < s.len:
     var x = 0
-    var failed = false
-    handleHexChar(s[i+1], x, failed)
-    handleHexChar(s[i+2], x, failed)
-    if not failed:
+    var success = true
+    success = handleHexChar(s[i+1], x)
+    success = handleHexChar(s[i+2], x)
```
Comment (timotheecour) [inline lib/std/private/decode_helpers.nim]: careful, that's a regression. Once again, use https://github.com/nim-lang/Nim/pull/16290/files#r543656396 and bug goes away

```
nim --eval:'import std/private/decode_helpers; var i=0; let c = decodePercent("%t9", i); echo (i,c)'
(0, '%')

nim --lib:lib --eval:'import std/private/decode_helpers; var i=0; let c = decodePercent("%t9", i); echo (i,c)'
(2, '\t')
```


=> this needs a test in tests/stdlib/tdecode_helpers.nim

**Review by dom96 — COMMENTED** [review]
If this is used so often maybe it should be placed in some public stdlib module instead of a new private one?

**Comment by dom96** [comment]
Also, as a side note, @timotheecour think you could add all your comments and then submit the review? From what I can see you're submitting each comment separately which really clutters the PR.

**Comment by timotheecour** [comment]
> If this is used so often maybe it should be placed in some public stdlib module instead of a new private one?

this can be done later though, right now this is just a refactoring; making it public would require more thought (eg the fact that `decodePercent` doesn't check the 1st char is `%` is not good for a public API)

> Also, as a side note, @timotheecour think you could add all your comments and then submit the review? From what I can see you're submitting each comment separately which really clutters the PR.

sure

File: tests/stdlib/tdecode_helpers.nim [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import std/private/decode_helpers
+
+
+block:
```
Comment (timotheecour) [inline tests/stdlib/tdecode_helpers.nim:4]: ok for this PR but in future I really prefer using the magic of templates to reduce boilerplate to 0:
```nim
template tdecodePercent(s): untyped =
  var i = 0
  let c = decodePercent(s, i)
  (i, c)
# then it's just 1 liners; furthermore it makes it easy if you want to edit
# the test as it's all in 1 place => DRY
doAssert tdecodePercent("%t9") == (0, '%')
doAssert tdecodePercent("19") == (0, '%')
...
```

(or even moving the doAssert inside, depending on your taste)

**Review by timotheecour — APPROVED** [review]

File: tests/stdlib/tdecode_helpers.nim (hunk 8 above)
Comment (ringabout) [reply tests/stdlib/tdecode_helpers.nim:4]: will do it in the future PR.

**Pull request merged**